synapse/server_notices/server_notices_manager.py

```diff
@@ -224,14 +224,27 @@ async def maybe_invite_user_to_room(self, user_id: str, room_id: str) -> None:
             if room.room_id == room_id:
                 return
 
+        user_id_obj = UserID.from_string(user_id)
         await self._room_member_handler.update_membership(
             requester=requester,
-            target=UserID.from_string(user_id),
+            target=user_id_obj,
             room_id=room_id,
             action="invite",
             ratelimit=False,
         )
 
+        if self._config.servernotices.server_notices_auto_join:
+            user_requester = create_requester(
+                user_id, authenticated_entity=self._server_name
+            )
+            await self._room_member_handler.update_membership(
+                requester=user_requester,
+                target=user_id_obj,
+                room_id=room_id,
+                action="join",
+                ratelimit=False,
+            )
+
```

Comment:
I was a bit alarmed by this, because it looks like `maybe_invite_user_to_room` is a generic function. But I see from its docstring that it is specifically intended for server notices.

I also wondered if having the initial membership be `invite` might be confusing. But the room's join rules presumably require invites, so I think the invite->join transition is inevitable.

On reflection: LGTM.